### PR TITLE
Allow setting of max instances for flowlets

### PR DESCRIPTION
**Pull request by ParthGandhi**
Allow the user to set maximum instances for any flowlets while building the spec. The default value is INT_MAX. 

For `InputFlowlets`, this value is restricted to 1.

### CODE REVIEW

**Comment by ParthGandhi** [comment]
A couple more changes to get this to work better:
1. Made the test handler thread-safe. Now the Netty test service need not be single threaded.
2. Making the `configure` method final in `InputFlowlet` means that the user can't set runnable specifications (i.e. failurePolicy, ResourceSpecification, and runtime arguments). So added a few methods that the user can use inside of `create` to do the same thing.

File: tigon-flow/src/test/java/co/cask/tigon/flow/test/BasicFlowTest.java [hunk 1]
```diff
@@ -133,39 +159,44 @@ public FlowSpecification configure() {
         .setName("testFlow")
         .setDescription("")
         .withFlowlets()
-        .add("generator", new GeneratorFlowlet(), 1)
-        .add("sink", new SinkFlowlet(), 1)
```
Comment (gokulavasan) [inline tigon-flow/src/test/java/co/cask/tigon/flow/test/BasicFlowTest.java:137]: What will happen if I set the number of SinkFlowlet instances here to 5 (even though the max instances in SinkFlowlet spec is set to 3) ?

Comment (ParthGandhi) [inline tigon-flow/src/test/java/co/cask/tigon/flow/test/BasicFlowTest.java:137]: Good catch. Added a similar check at deploy time, and a negative test case.

**Comment by ParthGandhi** [comment]
@gokulavasan Addressed comments. Please review.

File: tigon-sql/src/main/java/co/cask/tigon/sql/flowlet/AbstractInputFlowlet.java [hunk 2]
```diff
@@ -121,6 +141,42 @@ protected void addQuery(String sqlOutName, String gsql) {
...
+   */
+  protected void setFailurePolicy(FailurePolicy failurePolicy) {
+    this.failurePolicy = failurePolicy;
+  }
+
+  private FailurePolicy getFailurePolicy() {
+    return failurePolicy;
+  }
+
+  protected ResourceSpecification getResourceSpecification() {
```
Comment (gokulavasan) [inline tigon-sql/src/main/java/co/cask/tigon/sql/flowlet/AbstractInputFlowlet.java]: java doc?

**Comment by gokulavasan** [comment]
Small comment. Otherwise LGTM!

**Comment by ParthGandhi** [comment]
https://bamboo-prod.continuuity.com/browse/PRODRELALL-UNIT13-2

**Pull request merged**